Feature/vs logic refactor

lcoram opened the pull request:
I have some thoughts / questions about the design of the seed node... should we actually have nodes that contain 1 seed not 2, and then pass 2 seed nodes into the segmentation function. The question is in the long term how will we handle seeds? Will we want to keep a list of them (e.g. when seed nodes are added to scene we keep track of them in a list of seeds)? Or do we just want to poll the scene when calling the segmentation or split function and have the software ensure there is only ever one seed node (and if there is none do not call the function)?

lcoram committed:
changes to cmake to get it to compile with Slicer 4.8

lcoram committed:
ENH: removing usage of fiducial/markup, seedNode now uses double type, fixing tests based on changes

Refactoring logic so it no longer uses fiducial or markup, but instead seedNodes.
Removing fiducial/markup stuff from Displayable Manager and cleaning it up
Removing fiducial/markup stuff from Module Widget
Changing the SeedNode to use Double instead of Float
Fixes seed node test based on change to double
Tried to fix the segmentation test (but it still does not segment as far as the similarity image, so it fails the test)

lcoram committed:
TEST: changing similarity reference file, since currently unable to reproduce better segmentation output

lcoram committed:
TEST: changes to seed node to initialize to 0. Fixing tests.

lcoram committed:
FIX: fix warnings on linux test build.

RafaelPalomar reviewed (commented):
Style: 
- we don't generally want lines over 80 characters
- remove commented code lines which are not valuable
- avoid use of multiple calls to this->GetMRMLScene() in the same function. Use a vtkMRMLScene* scene variable.
- avoid the use of std::cout (vtkDebugMacro instead?)
- maybe replace asserts by checks+vtkErrorMacro (discuss)? 
-CallMergeLabelMaps is not a good name, how about just MergeLabelMaps?
- Separator between functions //-----------------------------
- avoid large if bodies. 

Checks:
 - Checking that mrmlscene is not null in functions using

Questions:
- why do we need SetActiveVolumeNode? Couldn't we just query it whenever needed?

TODO:
- fill out the PrintSelf functions

lcoram committed:
Removing the member variables about the active volume and accessing programmatically where needed.

lcoram committed:
DOC: adding comments in seedNode for functions created through vtkMacros

lcoram committed:
TEST: improving the logic unit test. Have good coverage of preprocessing and segment vessels.

lcoram committed:
TEST: changing logic to avoid repetition of code by creating a function to get the active volume. Then changing the test to have coverage of that and several other functions.

lcoram committed:
TEST: renaming and fixing the segmentation use case test.

lcoram committed:
ENH: adding functionality to logic to propagate volume and label selection. Adding to logic test for these functions.
Fixes #67

lcoram committed:
TEST: fix to set application logic in segmentation test.

lcoram commented:
In the logic now poll for the active volume when needed. Have tried to reduce the line length, and larger if bodies. Using debug macro for timing and a few other informational print outs. 

There are functions for propagating the volume selection (also used in the widget, for the dropdown), so this fixes the issue to do with that. 

Also, comprehensive error testing in the logic test, and also a good use case test for segmentation (renamed - for better naming convention).

RafaelPalomar reviewed (commented):
Very good! Some minor changes here.

LOGIC
- Documentation for AddSeedNode() function in vtk...Logic.h
- Documentation for GetPortalITKData() in vtk...Logic.h
- Documentation for GetHepaticITKData() in vtk...Logic.h
-  PrintSelf function in vtk...Logic.h is not correct. You should print the pointers even if they are NULL
- Separation mark between setmrmlsceneinternal and observemrmlscene
- if updatefrommrmlscene does not do anything there's no need for checking whether mrmlscene exists
- Pre-processImage: I don't see the check for mrmlscene pointer, yet you use the pointer.
- separation between mergelabelmaps and splitvesselsfromwidget.
- separation between splitvesselsfromwidget and splitvessels
- separation between splitvessels and updatemodels
- separation between updatemodels and reset3dview
- check all separations between functions.
- try to do your checks at the beginning of the function, the idea is to jump out of the function as soon as possible if a condition is not met.

NODE
- Complete the PrintSelf function

LOGIC TESTS
- Add tests for AddSeedNode
- Change the naming of the tests On...-->.... 
- Add test for SegmentVesselsFromWidget?
- Add test for SplitVesselsFromWidget?
- Add test for Update Models
- Add test for Reset3DView.
- AddTest for SetMRMLSceneInternal
- Check all functions that could have a test

lcoram committed:
ENH: Addressing comments from Rafael for pull request on branch VS_logic_refactor

Fixing formatting / spacers between functions. Completing printself functions in logic and seednode. Improving the testing of logic, by testing more functions.
